### DISCUSSION TITLE
AAA features important to me.

### DISCUSSION THREAD

**Discussion by decryptedchaos** [o3de/o3de#4703] [General]
I am heavily following the progress here,  i have dabbled in CryEngine all the way back to 3.x   Unreal 4.8-ish,  and Unity 2015?  

The problem i have with all of these engines,  is their overly complicated codebase,  Unreal while being a beautiful, photo realistic renderer the code-base is simply not idea for indie developers,   i have somewhat decent computer systems and i had servers which also shared compile time even then,  Unreal engine still took upwards of 18 hours to build at times.  

CryEngine was less of an animal,  but basically ZERO documentation,  accomplishing even the simplest task was little more than trial and error.  

Unity,  Ah honestly,  what has driven me away from Unity the most is the lack of high fidelity graphics, out of the box, not saying unity can't,  but to get it to produce a scene with the same fidelity as Unreal is quite a task.  and then there is C#   i know this one wont make me any friends,  but when i consider something like a game,  i feel that C# isn't leading itself well to optimizations  poor GC etc etc. 

As for lumberyard,   i never really messed with it,  i kind of wish i had now, so i had been more failure with the lineage of O3DE, my primary reasons for not using lumberyard was,  i didn't want to be stuck with AWS back-end,  that said i'm somewhat reconsidering that now.  

Getting to the point,  the title of this says "Features important to me"  

While some are of the opinion that graphic fidelity should be a secondary concern,  however i believe that graphical immersion is the first point of contact for a game,  making it feel as real as possible has never been more important. 

***I Should point out that i do want community discussion to take place on the points i've raised,  i want to talk with the devs,  i want to talk with the community using O3DE and see where everyone stands,  TLDR ; feel free to comment*** 

### Atom Renderer 

- Volumetric fog 
   - Deferred fog, doesn't seem to be enough
- Dynamic shadows 
   - Cascade through windows, trees etc.
- Sun shafts 
   - Controversial, i believe light shafts "God Rays" when used properly greatly add to the immersion of a scene

### Environment

- Dynamic day & night 
   - Sun position system,  would be nice to have,  but i could actually life without this one
   - This should lend itself to the dynamic shadows mentioned above.

- Water system 
   - Consider the new Water plugin for Unreal,   it's absolutely beautiful from a visual standpoint, 
   - Design wise it could use some work, the way it interacts with terrain is a bit finicky to say the least.
   - Impressive points.
      - Wave curve manipulation.
      - Water body blending.
      - Out of box water shader.
      - Spline based rivers

-  Nvidia Blast
   - This is very welcome feature, However the requirement of a Houdini licenses is not ideal for an indie developer
   - I haven't yet took a deep dive into blast and its requirements. 
      - Please clarify if this is a Nividia limitation or an arbitrary artificial limitation.
      - As games get more and more immersive, destructible environments become the norm.  
      - Being able to author these environments with open source tools would be ideal

- Height map Terrain System
   - I know this is somewhat being worked on.
   - Focus Points
      - Large Open World
      - Easy Height map import
      - Procedural height maps? (Just an idea, not of great importance to me)  
      - Height map deformation (Also would be a great feature to see, but not a show stopper)

### UI/UX
- Asset Browser
   - The Drill down menu is a bit awkward, drilling from C:\ --> all the way to project
- Dockable Windows
   - Docking is a bit buggy,  screen real-estate is going to be a premium very quickly.  

### Technical 

- Build Process
   - Streamline the build process
   - Currently it seems you must build the engine, and the project separately 
      - This could be my fault, but i must build the engine, and the open the project manager, build the project, then open the project via the project's exe, it won't open directly from the engine's project manager.  
   - Compile time isn't a big issue right now,  However it may be worth looking into implementing a distrbuted compile system I've used FastBuild and its open source so it would be worth considering.
- Scripting
   - I rather like the idea of using LUA for entity scripting, providing that:
      - We will/have proper c++ hooks exposed for actually doing useful things, 
      - Documentation is meticulous 
      - I would like to see a default skeleton script get generated automatically, just as a quality of life/ease of access feature.
  

**I will add more as it comes to mind**

### Documentation

It must go without saying that all the features in the world (CryEngine) are only as good as their documentation and ease of implementation. I hope keeping with the open source and community driven aspect of this new project, the documentation and training is kept relatively available for the community and not pay walled behind udemy courses and the like. 

I personally find that video documentation of individual systems,  and while this isn't really possible for extremely in-depth topics, a full library of videos explaining the architecture of how the engine functions would be extremely helpful in getting adoption, for example,   if i want to make a FPS  a video that went through all the ways to accomplish the basics for O3DE, as it differs from the other engines,  is going to be impressively valuable for indies 

I really can't stress enough how adequate documentation is a absolute requirement. 

To be more precise,  ecosystem training is what is key,  its how to setup the engine/project to be in-line with a goal,  and this is often the hardest part to figure out with an engine.  

For example,  If i want to make a certain style game,  and i need to accomplish this and this,  knowing the ecosystem of a engine as clear and straightforward as possible is very important.   as an indie,  i don't want to have to digest the entire C++ API just to figure out how to do something simple like setup a score system.  or setup teams.   

Things that are framework type ideas for games: 
- For FPS
   - You need a FP Controller
   - You Need Weapon attach points
   - You Need Teams
   - You Need Score
   - You Need Objectives
- For RPG
   - You Need Skills
   - You Need Multi-type weapons
   - You Need NPCs 
   - You Need AOE effects
   - Quest or Objective systems.  
   
I'm not technically requesting that all the work be done for us,  what i am saying is  there are concepts that are common to specific genre, having a clear vision for how these things "Should" be accomplished in a new engine from the start is incredibly valuable,  like a best practice guide,  for getting the ball rolling for a specific genre.  

This is also why i would prefer, highly detailed walk throughs to get to a starting point over example projects,   because example projects you have to reverse engineer,  and hack and slash at things to get what you want out of them,   it would be far more beneficial to learn how to build it from scratch in a very detailed walk through vs being handed a sample project. 

**To be continued/updated as needed**

**Comment by amznestebanpapp** [comment]
> Currently it seems you must build the engine, and the project separately
> This could be my fault, but i must build the engine, and the open the project manager, build the project, then open the project via the project's exe, it won't open directly from the engine's project manager.

Is not necessary. There are 3 main workflows:
1. engine-centric: compiled from the engine point of view and optionally include your project. This will build the engine+project. Practically speaking, you can think of this as the targets provided by the engine and the targets provided by the project added to your solution. 
This workflow has been maintained because it allows to build the engine without a project (e.g. if you are just running unit tests); it also allows to build multiple projects at once. It was also the only workflow in Lumberyard, so it was supported since the transition to CMake.
2. project-centric / engine-source: this is very similar to (1) except that the entry point to CMake is the project. Similar to (1) you get all the targets from the engine and project. 
This workflow has the benefit of allowing to switch the engine you are using, even switch between (2) and (3). This workflow is the desired one if you are modifying the project and engine or you prefer to have the engine's source to be able to debug/learn about it as you develop the game (this would be my recommendation to your case).
3. project-centric / engine-prebuilt: this is what seems you are using. The prebuilt part doesn't need to be done by you. There is an installer that can give you that "engine-prebuilt" part. The intention of this workflow is for those that don't need to dive into the engine source or require to do modifications. 

The purpose of documenting and talking about building the engine (for using it in (3)) is because that will likely be the default in a team. The expectation is that a team building a project will have some engine modifications that they will want to distribute to the rest of the team. An "Engine engineer" will modify the engine and through CI will provide that modified engine in a prebuilt form to a "Project Engineer" that will build the project against it (maybe even providing a precompiled version to Artists to work on levels/assets). 

My recommendation to use (2) for your case is because it looks like you are interested in learning about the engine, being up-to-date, provide contributions, debug it, etc. All while building your project. Thats what (2) provides.
You can switch between (2) and (3) with prebuilt versions done by the installer, that will reduce your build times, however, will limit your ability to modify/debug.

**Reply by decryptedchaos** [reply]
I agree with your assessment,  The part that was unclear about this when i setup the project is. 

With project-centric / engine-source,  are changes to the engine-source  automatically inherited to project without re-building engine-source.   e.g  when i run cmake --build ...  in the project dir are any changes/pulls from the engine-source repo/folder automatically applied to what's being compiled into project..  

This was unclear to me,  but from your description that is what i believe to be the case.

**Reply by amznestebanpapp** [reply]
> With project-centric / engine-source, are changes to the engine-source automatically inherited to project without re-building engine-source. e.g when i run cmake --build ... in the project dir are any changes/pulls from the engine-source repo/folder automatically applied to what's being compiled into project..

Yes. 
Practically speaking, the project is including the engine source: the project cmake script includes `<enginepath>/cmake/Findo3de.cmake` that ends up including `<enginepath>/CMakeLIsts.txt` and including all the cmake scripts and files from the engine source.
So, when you run `cmake --build ...` in the project dir, you are effectively building the build targets provided by the engine.
This is very evident to see if you open the solution generated in VS, you will see all the targets from the engine source there as well. If yo do project-centric/engine-prebuilt, you will only see the project targets.

**Comment by amznestebanpapp** [comment]
> Compile time isn't a big issue right now, However it may be worth looking into implementing a distrbuted compile system I've used FastBuild and its open source so it would be worth considering.

This is a really long discussion. Bare with me. I am likely going to miss a lot of important details, feel free to ask more and I will do my best.

It does sound tempting, and to be honest, it was my choice before converting WAF to CMake. 
At the time the conversion started (1.5-2y ago), FastBuild did not support all platforms and had a lot of issues generating native projects. That has gotten better over the last 2 years, however, there are fundamental differences that even if FastBuild worked well, would still cause issues.

There are multiple aspects to a build system, here I am focusing on 3: 
1. how fast it builds 
2. how good developer experience it provides
3. how easy/possible it is to extend/script

FastBuild does (1) very well. Although we saw some issues on scaling the distributed cache and distributed compilation, I think those are likely already fixed or fixable.
Arguably, this is where FastBuild > CMake. However, note that its not apples to apples comparison. CMake **does not build**. So the comparison is more like FastBuild > MSBuild or FastBuild > Ninja.

For (2) the issues start to be more evident when you are developing (vs building). The way that FastBuild integrates in an IDE is by generating build scripts that call back to FastBuild. So, for example, in the case of Visual Studio, it will generate `vcxproj` files that will have `CustomBuild` rules that will invoke FastBuild back to perform every build action.

This creates some issues:
- building a compilation unit (a cpp) will call FastBuild again. In a big project, parsing all the build scripts to build one cpp file ends up being 20-40s just of processing time to spend 1-5s of building. We had this same problem in WAF (Lumberyard). The way WAF was integrating in Visual Studio was exactly the same. That caused 20-40s of WAF parsing/processing time to end up spending 1-5s of building.
FastBuild has a caching system that causes reparsing when the files are dirty. The tracking of all the files that describe the build will still be some time taken. Since we would likely not be able to describe the full build in FastBuild, we may also need to pay here a roundtrip time to some other tool to detect other changes and make sure that we are not re-generating FastBuild if not necessary. Basically, imagine we do SharpMake->FastBuild->vcxproj, we then build a cpp and have to do VS->FastBuild->SharpMake->FastBuild to perform the build. 
In CMake, this does not happen because CMake generates a vcxproj that is built by MSBuild (which is what Visual Studio uses to build under the hood). When you go and compile one compilation unit (one cpp) is a direct call to `CL.exe` with the required parameters. This means that you dont get those 20-40s of parsing all the build scripts, just the 1-5s required to build that cpp.

- A related problem is that the settings you see in the IDE are not used. FastBuild (and similar what we did with WAF) just generate a callback to FastBuild (and potential parameters baked into a command line invocation). FastBuild generates a type of project (Make) that just hooks every file back to FastBuild. Parameters, include paths, compile definitions, etc are not passed. This means that if you change a compile flag in e.g. Visual Studio, you wont see anything happen in the build. You will see it building because MSBuild will consider the whole project dirty, but then FastBuild will consider it clean. The result is that none of the settings in the IDE are usable. 
In CMake, this does not happen, you can change the settings and see those changes reflected. I use this a lot of fixing build problems. For example, I add a `/showIncludes` to a compilation unit/project and build that. I can diagnostic and experiment with settings and see those settings reflected. If I want to nuke my changes, I regenerate, if I want to preserve those changes, I have to move them to CMake (it would be great that modifying the properties in VS would modify the cmake scripts directly, but that is an impossible ask). 

- Another related problem is how well is FastBuild generating the settings that MSBuild expects. This one is really tricky. Take the "[external header](https://docs.microsoft.com/en-us/cpp/build/reference/external-external-headers-diagnostics?view=msvc-160)" feature taken out of experimental in 16.10.
We are using this to disable warnings coming from 3rdParty and enabling higher warning levels in O3DE's code. Without these flags it would be very hard to implement higher warning levels than the ones done by the 3rdParty. O3DE code is now at W4 + extra warnings enabled (as errors). Lots of the headers in 3rdParty are not at that level. We did some partial implementation without the flags, you can see a lot of Qt headers wrapped in push/pop warning disable macros. However, without the flags, it would have been a way-bigger amount of work.
For these flags to work, we need to pass to the compiler `/external:W0` and then pass the include paths for 3rdParty through `/external:I`. However, when you generate the `vcxproj`, you need to pass an xml entry that looks like: `<ExternalWarningLevel>TurnOffAllWarnings</ExternalWarningLevel>`
I may be wrong, but I also think that we wouldnt be able to implement that functionality without modifying FastBuild itself. I dont see FastBuild providing a way to pass things to the generator directly, and it also doesnt seem to provide those options. The projects it generates look like a "file container" and not a proper VS project. I also dont see FastBuild supporting transference of properties between targets. Likely we would have to hide the paths behind a variable and make every target not just depend on the 3rdParty but also add the compilation options to the targets themselves. This could end up in a spaghetti of places, specially when those headers are used in other headers.
In CMake's case, they already added support for it, but when we wanted it, we were able to integrate it manually, without modifying CMake.

- Incremental linking/edit-continue/hot reloading: all these are settings that are mostly driven by Visual Studio. For Visual Studio to understand that it can do edit-continue, it has to be able to inject special parameters to the compilation of a cpp. These features work out of the box with CMake because its all dealt within Visual Studio, FastBuild would have to understand how to hook itself up in that scenario, not impossible, but since is a "middle-man" in the build, it needs to handle it. And it looks like it doesnt ATM.

- IDE integration. Some IDEs already support CMake as a "project data model". For example, in Visual Studio, you can open the CMake scripts directly instead of generating `vcxproj` files. Same doesnt happen with FastBuild.

- I know I have focused a lot on VisualStudio, however, similar things happen with e.g. XCode and Android Studio.

Regarding (3): FastBuild is both a "build description language" (how you declare build targets) and a "build system" (the one that performs the actual build). 
It behaves really well as a "build system", but it lacks a lot on "build description language". Specifically, when you start to try to make scripts for multiple platforms, compiler and version auto-detection, SDK, etc you start to see how difficult it is to describe everything. If you consider the amount of features we had in Lumberyard/WAF, some of those things get really difficult to define in FastBuild. FastBuild is also really barebones, it doesnt bring a lot of that scripting already done for you. If you look at cmake's `share\cmake-3.21\Modules` folder, you will see a lot of toolchain/IDE bootstrap logic and tooling. Things like `share\cmake-3.21\Modules\Platform\Android` takes a lot of the burden to setup that platform.
I personally see FastBuild closer to Ninja, where "the build files are human-readable but not especially convenient to write by hand". Most usages of FastBuild in big projects I have seen are generating FastBuild (i.e. the build is authored in some project-generator that generates FastBuild, like SharpMake). We couldnt take that path either because SharpMake is not supported in all platforms (there is no support of it running on Mono). And CMake does not generate FastBuild...

CMake provides testing/install/packaging. The amount of scripting possible in CMake is huge in comparison. Just compare the documentation and you will see how far CMake is in comparison.
We were able to build a lot of features into the build scripts, things like runtime dependencies, platform abstraction, plugin loading description, install layout, packaging, monolithic builds, etc. 

We also considered things like community support, likeliness for a developer to know about it, etc. Just doing a couple of searches to solve some issue in FastBuild vs CMake you will see how different it is the amount of information/support out there. We had this problem with WAF, it was really hard for developers to fix things because WAF is not widely used. With CMake we have found tons of examples, questions in CMake's issue tracking, etc. 
If things go south, we can also talk to/hire Kitware to get help.

My hope for FastBuild/CMake is that CMake eventually generates FastBuild. It was tried in a branch but never made it back to main:
https://gitlab.kitware.com/cmake/cmake/issues/15294
https://github.com/InSimo/CMake/commits/fastbuild-attempt4
That is something maybe worth to push for if you really want to use FastBuild.

There are also other options to do build distribution, you can use anything that is at the command invocation level, e.g. distcc. That integrates easily to CMake. Maybe you can use FastBuild that way. It would be great if FastBuild provided the distribution/caching at command-line invocation level. Since its open source, thats maybe a path you could be willing to take. Basically I think taking the piece of FastBuild that does the distribution/caching and make it a tool like distcc will allow to integrate it directly into other places. 

You can also use anything that understands/converts any of the generated build systems that CMake supports, for example, you can use [this](https://github.com/liamkf/msfastbuild) to generate FastBuild from the generated "Visual Studio" projects. Or you can use something like IncrediBuild directly on the Visual Studio projects. 

I honestly would not recommend doing any of these things. The issue you will find is after the first build. When working in small changes, incremental builds will not work as expected and/or will deteriorate a lot of the IDE experience. We experienced this a lot, we would use some distribution build tool, then when we want to build 1 cpp, it will take a long time to go through that or completely invalidate the build at unexpected times. 

CMake does a great job at generating proper Visual Studio/XCode/Ninja/etc projects, which integrate properly into the IDEs. Over time, you will likely note that you dont do big builds but rather smaller builds with your changes. Once you dive more and more in the source code, you will find yourself building just one target, just one test suite, just one cpp to validate your changes. It will become less frequent building "everything" and definitely less frequent to be in situations where a lot has to be built to get the Editor/GameLauncher.
I understand this is not going to be your experience at the beginning. At the beginning you will be consuming a lot of big changes externally until you feel confident with the source code and are able to branch, make changes and work on your changes. This model somehow is benefiting the latter group, however note that we have been working on providing pre-compiled versions for the first group.

The problem may also be exacerbated now becasue we dont have frequent stabilization branches. You are likely receiving a lot of changes daily. Eventually we will have stabilization branches that you will be based of and just receive smaller bug fixes, and just deal with big changes when switching stabilization versions. My recommendation until we are at that point is to take one commit point and work out of that, moving that commit point less frequently (e.g. weekly) and dealing with the long-build overnight. Or, if you are not interested in the engine's code and debugging it, to use the nightly installer. 
That would keep you working on the project and a modified less-frequently codebase of the engine where you should not be getting major changes, making the big-build scenario less of an issue.

**Reply by decryptedchaos** [reply]
First,  I applaud and welcome the very detailed explanation,  to me,  this is the only way one can learn the system is pose questions/suggestions. And now what had seemed like an obvious solutions,  I now know why it isn't as simple as meets the eye.

My experience with FastBuild was born of messing with Unreal Engine,  it's so massive,  and my build/dev system isn't a slouch it's a 2800x @ 4Ghz 32GB Ram and a 3200mb/s NVme  and skipping over the hour long rant i have about how many time i had to clear stuff off my NVme to hold UEs 250GB source folder.   The compile times were bad to say the least. 

I had some days where it would simply spend 15 hours compiling, then fail due to some platform dependency i wasn't even using.   I got so fed up with it,  because i was building 3-4 days out of the week,  just constant compiling,  hardly any work getting done.  And when it finally did build after 20+ hours..  another update with a bug fix that i needed.... 

So i dug around and found FastBuild which up until recently didn't exactly work out of the box on Unreal Either,  code from here and there,  and a few tweaks and i finally made it work 
> One should note that,  you are correct,  if a user was to try to google  FastBuild error:   They were more likely to find buried treasure. 

Anyway,  I set it up,  because it was getting out of hand,   and i have a few Xeon servers locally in a rack.   But with that said,  i experienced exactly what you described,  long pauses where it appeared to be just sitting there,  "enjoying a cold one"  I didn't pay it a great deal of attention it did seem like it was "Overall" a little faster,   but i did also experience strange issues where the cache would become invalidated and required clearing to build.   

I can completely understand the tooling nightmare it might be to adopt something like this,  and I agree with your assessment of its usefulness at this point,  

> I guess moral of the story,  is  I just hope it doesn't become as bad as Unreal 20+ hour compile times

**Reply by amznestebanpapp** [reply]
We also have some more cleanup ahead of us. GridMate is going to be replaced by AzNetworking and we still need to fully remove CryCommon and CrySystem. There have also been some PRs from the community, specially from @nemerle to cleanup and reduce build times. 

Keeping built times manageable in a codebase of this size requires continuous cleaning. Sadly, a lot of C++ performance and game techniques are not built-time friendly. There is also the problem that a new system is introduced and the old one is left for backwards compatibility. It's going to be a continuous fight to have a balance.

It should (hopefully) never get to 20+hour compile times. What is your current compilation time?
We also hope that a nightly installer will help to avoid the compilation of the engine. Is that something you are using?

**Reply by decryptedchaos** [reply]
Good points here,  Oh my build time with O3DE right now is probably only 20-40 minutes,  so its no big issue,  even 1, 2, 3 hours is acceptable. 

I just don't want to be in compile/dependency misery like i've experienced in Unreal,  being completely aware huge code bases naturally take a lot to process/compile.  Unreal was always just problem after problem after problem. most of which were derived from some third party platform support module, that wasn't even relevant to 80% of the users.

**Comment by amznestebanpapp** [comment]
> For example, If i want to make a certain style game, and i need to accomplish this and this, knowing the ecosystem of a engine as clear and straightforward as possible is very important. as an indie, i don't want to have to digest the entire C++ API just to figure out how to do something simple like setup a score system. or setup teams.

This is a great idea, and I would suggest that you push this as a proposal to the community.
We have a template system that hooks up into the Project Configurator (o3de.exe) and should be able to predefine templates for those kind of projects. Take a look at the `Templates` folder, that is where the "DefaultProject" and "MinimalProject" are defined. There is nothing preventing someone from adding templates for those cases. It likely will need some agreement from the community on how far/complete to take the template. You could have a template there that is a hello-world demo of each genre.

> I'm not technically requesting that all the work be done for us, what i am saying is there are concepts that are common to specific genre, having a clear vision for how these things "Should" be accomplished in a new engine from the start is incredibly valuable, like a best practice guide, for getting the ball rolling for a specific genre.

Likely the point to start would be in the `games` channel in discord (under INDUSTRY VERTICALS). I would expect you to find others to help and then a RFC with the actual proposal. Then making the PR and getting the things in. As a first pass I would suggest to create this genres under the "Games" vertical, meaning, separate those templates so we let other verticals to also add templates.
The templates can be barebones at the beginning and can evolve over time.
This would have been great for the o3de game jam. It would be great to have it as a starting point for everyone to build their games. Maybe it can be done for the next one.

> This is also why i would prefer, highly detailed walk throughs to get to a starting point over example projects, because example projects you have to reverse engineer, and hack and slash at things to get what you want out of them, it would be far more beneficial to learn how to build it from scratch in a very detailed walk through vs being handed a sample project.

Idem to above, I would expect the community to help with this. The o3de.org website is also in a [repository](https://github.com/o3de/o3de.org). You should be able today to create an entry like that and get it in. Likely the rest of maintainers/contributors from that repo will want to validate that the steps work and the wording/etc is proper. 
If you need to add templates for it, you can also create a PR against [o3de](https://github.com/o3de/o3de) and get things in.

All this may be too much work for you or for one person to take. I would suggest to start a conversation in the game channel and see if there are others willing to help. I am pretty sure you will find others willing.
Likely you can create these things while you dive into the engine and discover that "starting point", once you have the "starting point" is a matter of converting it to a template, document it and creating some more documentation for others to understand it and know its there.

**Reply by decryptedchaos** [reply]
> We have a template system that hooks up into the Project Configurator (o3de.exe) and should be able to predefine templates for those kind of projects. Take a look at the Templates folder, that is where the "DefaultProject" and "MinimalProject" are defined. 

This sounds very promising actually,  a few ideas have already come to mind on how to keep such a thing point focused.  

One thing I've seen with other "sample projects" or templates,  is they either don't go far enough,  leaving the user still struggling to figure out what the next step should be or..  For something like CE GameSDK  they went way too far,  i remember trying to reverse engineer GameSDK and it was a nightmare,  when i just wanted the skeleton,  quickly learned that i can't just cut this and this out  without creating all kinds of problems.

**Reply by amznestebanpapp** [reply]
Something that could be done is a more "incremental" template. A simple project base template could be used and then gem templates can add that extra functionality as add ons to the project.
You could have a "Game" template like a "FPS camera" that a user can add that will take care of setting up the camera. Similar with other pieces like "FPS character animation"; etc. 
Gems are plugins to both the engine and the project, there is no limit on the amount of plugins a project can have, if you organize it that way, isolated enough, they can be modular and pluggable.
You likely are going to find some practical limitations, if you want a FPS camera but be able to switch to 3rd person view you may need a different gem. Or you may find it easier to add it to the same gem and have something like a compile switch or variable define if it is enabled or not. Striking the balance of simplicity vs features is going to be tricky. I would suggest to have a "FPS camera simple" for someone to start with and a "FPS camera advanced" for more advanced features.

**Comment by micronAMZN** [comment]
Picking up the discussion here from the docs issue filed on O3DE.org.

In regard to features in general, there is a [roadmap for O3DE here](https://github.com/o3de/o3de/projects/1), where you can see what is being developed. It is not the sum total of development efforts around O3DE, just what has a public commitment through the SIGs. Some of it may not be terribly exciting from a creator perspective at the moment, but keep an eye on it :)

Blast is a feature that is very important to me. The docs are currently out of date, and you might be able to use any DCC to create Blast assets. You'll have to create a fracture hierarchy in your application of choice, export it as an FBX asset. and try to use NVIDIA's tools to create the blast asset. I hope to provide documentation and a better solution in the future. I have it on my personal priority list to sort this out in the next couple months.

My overall goal is that from a tutorial standpoint is that we don't cater to any particular commercial DCC application, and that from a source perspective, we support other open source efforts such as Blender. We can't guarantee this will always be the case, though. Sometimes O3DE may need to leverage the particular features of certain applications for commercial developers.

While they aren't well documented and still in progress, do try out the terrain tools in development. I think they already might fulfill some of your requests. There should be some talks available in the near future from O3DE Con that can help. Now is the best time to provide feedback on them.

On documentation...

The first thing, and this is personal opinion, is that I think it creates incredible overhead, and could be quite confusing for most users to document O3DE from the perspective of how it works compared to other engines. It's much more effective to write documents assuming some level of _general_ knowledge about games, graphics, and simulation. "No prior experience at all" is a valid level of knowledge. Some documentation will cater to new users, some docs will have to assume some level of advanced understanding. I do think it might be helpful to create some sort of cheat-sheet like documentation as an appendix to help users make the mental shift from other engines to O3DE. That will have to be done by a contributor who understands the other engine and O3DE well enough to make the comparison, though. It would also have to be constantly updated as both O3DE and other engines advance.

Beyond that, I think most of us involved in documentation and tutorials agree with you. We are trying to provide both written and video components to tutorials as much as possible. We are shifting heavily toward focusing on tutorials, so specific tutorial requests are very welcome. Having concise tutorials that document a specific game system can be most effective. Many of these concepts are functionally identical across game genres, and if you have enough of them, you can combine them into complete series for specific game types. I also like starter projects, but only as templates for game types. Going through the effort of fully developing an example project, something close to a vertical slice for a game as you noted, is too difficult and time consuming for most users to unravel and learn anything.

**Reply by decryptedchaos** [reply]
In Regards to blast,  that's good to hear,   I've been going over both the O3DE docs and the NVIDIA docs for Blast.  

It kind of a wasteland,  i've found some references to a tool called the "Blast Authoring Tool" seen below. 

However, i can't find any further reference to this tool,  the NVIDIA docs simply say  **Download Link Needed** The Blast github repo has a AuthoringTool.x64.exe  when you compile it,  However it doesn't appear to be the same tool,  no GUI,  it's a console app for converting FBX assets into blast assets. 

But this isn't the solution,  because we need to understand how to setup the fractures on the mesh,  seemingly the tool they include in the repo assumes this has already been done,  and the rest of their docs are about using the tool shown below ~~which i can't find anywhere.~~ Actually i did find it,  you have to download the old version (1.15)

![Blast Tool](https://gameworksdocs.nvidia.com/Blast/1.1/authoring_docs/_images/BLAT_EditorBase.jpg)

**Reply by micronAMZN** [reply]
I went through the same thing a while back and filed an issue about it. I couldn't find that tool either. I found it tonight thanks to you, but can't get it to export a `.blast` asset. I can fracture an `.fbx`, the physics work in the tool. It says it's writing a `.blast` file, but it doesn't.

You can fracture meshes and play with Blast in Omniverse Create, but there's no way to get a `.blast` asset out of Omniverse Create that I can find. The NVIDIA command line tool (and that GUI tool) are really only good for making test data. I spent the evening here trying to get Blast to work. My advice, don't spend any time thinking about it for the moment. [See this issue](
https://github.com/o3de/o3de/issues/2617).

Fracturing the mesh is the easy part. You can do that in any application. All you need to create is an `.fbx` file with a hierarchy of meshes. You can do this in Blender, Maya, Houdini, and Omniverse Create without any additional tools.

The issue with Blast in O3DE that I see is that, there is no way to create a `.blast` asset from a user defined fractured mesh. Even if you do have a `.blast` asset, it seems Asset Processor won't process it. If the Asset Builder did work, it would produce a slice, which may no longer be supported by O3DE.

When the issue was closed a few days ago, I assumed that I could write a Python Asset Builder to sort everything out, but there are some significant problems with Blast at the moment. It needs some real developer love.

**Reply by decryptedchaos** [reply]
> I found it tonight thanks to you, but can't get it to export a .blast asset. I can fracture an .fbx, the physics work in the tool. It says it's writing a .blast file, but it doesn't.

For reference,  i was able to get it to export .blast   see below for settings   The asset processor seems to at least accept it,  no in-depth testing in engine yet.   It does appear buggy though,   the .blast appears to be binary too,  not exactly easy to troubleshoot. 

Moreover, It appears that `Embed collision in fbx` causes problems with the asset processor, while we likely need this data, turning it off does at the very least allow the assets to be exported and ingested.

![blast-export](https://user-images.githubusercontent.com/1302391/137624695-598a65c2-640f-4009-bf22-4da63de13222.jpg)

![blast-processing](https://user-images.githubusercontent.com/1302391/137624698-fc7b3f61-0839-4e06-b0ab-cb607ec291bc.jpg)
